File: bidirectional_voice_streaming_client.py
import asyncio
import websockets
import os
import base64
import wave
import sys
import json
from io import BytesIO

# Arguments
recording_path = sys.argv[1] if len(sys.argv) > 1 else '/tmp/py_audio.wav'
port = int(sys.argv[2]) if len(sys.argv) > 2 and sys.argv[2].isdigit() else 3001

# Log details
print(f"Listening on port {port}, writing incoming audio to WAV file {recording_path}")

# WebSocket Server
async def handle_connection(websocket):
    print(f"Received connection from {websocket.remote_address[0]}")
    
    # WAV file parameters (assumptions based on your code)
    sample_rate = 8000  # The sample rate (Hz)
    num_channels = 1    # Assuming mono audio
    sample_width = 2    # 16-bit PCM (2 bytes per sample)
    frame_rate = sample_rate

    # Create a wave file for writing audio
    wstream = wave.open(recording_path, 'wb')
    wstream.setnchannels(num_channels)
    wstream.setsampwidth(sample_width)
    wstream.setframerate(sample_rate)

    try:
        async for message in websocket:
            if isinstance(message, bytes):
                print(f"Buffer size: {len(message)}")
                wstream.writeframes(message)  # Write raw PCM audio data to the WAV file
            else:
                # Read and encode the WAV file
                wav_path = '/usr/local/freeswitch/sounds/en/us/callie/ivr/8000/ivr-welcome.wav'
                if os.path.exists(wav_path):
                    with open(wav_path, 'rb') as f:
                        wav_buffer = f.read()
                        base64_data = base64.b64encode(wav_buffer).decode('utf-8')
                        response = {
                            "type": "playAudio",
                            "data": {
                                "audioContentType": "wav",
                                "sampleRate": 8000,
                                "audioContent": base64_data,
                                "textContent": "Hi there! How can we help?"
                            }
                        }
                        await websocket.send(json.dumps(response))
                else:
                    print(f"Error: WAV file not found at {wav_path}")
                    
    except websockets.exceptions.ConnectionClosed as e:
        print(f"Socket closed {e.code}:{e.reason}")
    finally:
        wstream.close()

# Start the WebSocket server
async def main():
    server = await websockets.serve(handle_connection, "localhost", port)
    print(f"WebSocket server started on port {port}")
    await server.wait_closed()

# Run the server
if __name__ == "__main__":
    asyncio.run(main())

File: bidirectional_voice_streaming_client.js
const WebSocket = require('ws');
const fs = require('fs');
const argv = require('minimist')(process.argv.slice(2));
const recordingPath = argv._.length ? argv._[0] : '/tmp/audio.raw';
const port = argv.port && parseInt(argv.port) ? parseInt(argv.port) : 3001
let wstream;

console.log(`listening on port ${port}, writing incoming raw audio to file ${recordingPath}`);

const wss = new WebSocket.Server({
    port
});

wss.on('connection', (ws, req) => {
    console.log(`received connection from ${req.connection.remoteAddress}`);
    wstream = fs.createWriteStream(recordingPath);

    ws.on('message', (message) => {
        
        if (typeof message === 'string') {
            console.log(`received message: ${message}`);
        } else if (message instanceof Buffer && message.length==320) {
            console.log(`Buffer size: ${message.length}`);
            wstream.write(message);
        }
	else{
	    console.log(typeof message);
            const wavPath = '/usr/local/freeswitch/sounds/en/us/callie/ivr/8000/ivr-welcome.wav';
            const wavBuffer = fs.readFileSync(wavPath);
	    const base64Data = wavBuffer.toString('base64');
            response = {
		"type": "playAudio",
		"data": {
			"audioContentType": "wav",
			"sampleRate": 8000,
			"audioContent": base64Data,
			"textContent": "Hi there!  How can we help?"
			}
		}
	    const data = JSON.stringify(response);
            ws.send(data);
        }
    });

    ws.on('close', (code, reason) => {
        console.log(`socket closed ${code}:${reason}`);
        wstream.end();
    });
});
